File: HW4/src/optim.rs
use nalgebra::DMatrix;
use simple_error::SimpleError;

pub struct OptimResult {
    pub x: DMatrix<f64>,
    pub iter: usize,
    pub converged: bool,
}

pub fn newton_w_equal(
    f: fn(&DMatrix<f64>) -> f64,
    grad: fn(&DMatrix<f64>) -> DMatrix<f64>,
    hess: fn(&DMatrix<f64>) -> DMatrix<f64>,
    x0: DMatrix<f64>,
    A: DMatrix<f64>,
    b: DMatrix<f64>,
    eps: f64,
    max_iter: usize,
) -> OptimResult {
    let mut x = x0.clone();
    let init_resid = &A * &x - b;
    if init_resid.norm() > 0.0001 {
        panic!("Invalid initial guess");
    }
    let r: DMatrix<f64> = DMatrix::from_vec(0, 0, vec![]);
    let b: DMatrix<f64> = DMatrix::repeat(A.shape().0, 1, 0.0);
    for iter in 0..max_iter {
        let step = match min_quad_w_equal(&hess(&x), &grad(&x), &r, &A, &b) {
            Err(e) => {
                break;
            }
            Ok(result) => result,
        };
        let dec = newton_dec(&step, &A);
        if f64::powf(dec, 2.0) / 2.0 < eps {
            return OptimResult {
                x: x,
                iter: iter,
                converged: true,
            };
        }
        x += step;
    }
    return OptimResult {
        x: x,
        iter: max_iter,
        converged: false,
    };
}

pub fn newton_dec(h: &DMatrix<f64>, A: &DMatrix<f64>) -> f64 {
    if h.shape().0 != A.shape().0 || A.shape().1 != h.shape().0 {
        panic!("Arguments unconformable");
    }
    let prod = h.transpose() * A * h;
    return prod[(0, 0)];
}

pub fn min_quad_w_equal(
    P: &DMatrix<f64>,
    q: &DMatrix<f64>,
    r: &DMatrix<f64>,
    A: &DMatrix<f64>,
    b: &DMatrix<f64>,
) -> Result<DMatrix<f64>, SimpleError> {
    let P_dims = P.shape();
    let q_dims = q.shape();
    let A_dims = A.shape();
    let b_dims = b.shape();
    if P_dims.0 + A_dims.0 != q_dims.0 + b_dims.0 || q_dims.1 != 1 || b_dims.1 != 1 {
        panic!("Arguments unconformable");
    } else {
        let lhs: DMatrix<f64> = min_quad_w_equal_lhs(P, A);
        let rhs: DMatrix<f64> = min_quad_w_equal_rhs(q, b);
        let decomp = &lhs.lu();
        let sol = match decomp.solve(&rhs) {
            None => return Err(SimpleError::new("Could not compute LU decomp")),
            Some(decomp) => decomp,
        };
        let result = min_quad_w_equal_result(sol, A_dims.1);
        return Ok(result);
    }
}

fn min_quad_w_equal_lhs(P: &DMatrix<f64>, A: &DMatrix<f64>) -> DMatrix<f64> {
    let P_dims = P.shape();
    let A_dims = A.shape();
    let mut lhs: DMatrix<f64> = DMatrix::<f64>::zeros(P_dims.0 + A_dims.0, P_dims.1 + A_dims.0);
    for i in 0..P.nrows() {
        for j in 0..P.ncols() {
            lhs[(i, j)] = P[(i, j)];
        }
    }
    for i in 0..A.nrows() {
        for j in 0..A.ncols() {
            lhs[(P.nrows() + i, j)] = A[(i, j)];
            lhs[(j, P.ncols() + i)] = A[(i, j)];
        }
    }
    return lhs;
}

fn min_quad_w_equal_rhs(q: &DMatrix<f64>, b: &DMatrix<f64>) -> DMatrix<f64> {
    let mut rhs: DMatrix<f64> = DMatrix::<f64>::zeros(q.nrows() + b.nrows(), 1);
    for i in 0..q.nrows() {
        rhs[(i, 0)] = -q[(i, 0)];
    }
    for i in 0..b.nrows() {
        rhs[(q.nrows() + i, 0)] = b[(i, 0)];
    }
    return rhs;
}

fn min_quad_w_equal_result(sol: DMatrix<f64>, nrows: usize) -> DMatrix<f64> {
    let mut result: DMatrix<f64> = DMatrix::<f64>::zeros(nrows, 1);
    for i in 0..nrows {
        result[(i, 0)] = sol[(i, 0)];
    }
    return result;
}

#[test]
fn dec() {
    let h: DMatrix<f64> = DMatrix::from_vec(2, 1, vec![1.0, 2.0]);
    let A: DMatrix<f64> = DMatrix::from_vec(2, 2, vec![3.0, 5.0, 4.0, 6.0]);
    let result = newton_dec(&h, &A);
    let expected = 45.0;
    assert_eq!(expected, result);
}

#[test]
fn min_univar_quad_w_equal() {
    let P: DMatrix<f64> = DMatrix::from_vec(1, 1, vec![2.0]);
    let q: DMatrix<f64> = DMatrix::from_vec(1, 1, vec![-8.0]);
    let r: DMatrix<f64> = DMatrix::from_vec(1, 1, vec![16.0]);
    let A: DMatrix<f64> = DMatrix::from_vec(1, 1, vec![3.0]);
    let b: DMatrix<f64> = DMatrix::from_vec(1, 1, vec![5.0]);
    let result = min_quad_w_equal(&P, &q, &r, &A, &b);
    let x = result.unwrap()[(0, 0)];
    assert_eq!(x, 5.0 / 3.0);
}

#[test]
fn min_multivar_quad_w_equal() {
    let P: DMatrix<f64> = DMatrix::from_vec(2, 2, vec![2.0, 0.0, 0.0, 2.0]);
    let q: DMatrix<f64> = DMatrix::from_vec(2, 1, vec![0.0, 0.0]);
    let r: DMatrix<f64> = DMatrix::from_vec(1, 1, vec![0.0]);
    let A: DMatrix<f64> = DMatrix::from_vec(1, 2, vec![-1.0, 1.0]);
    let b: DMatrix<f64> = DMatrix::from_vec(1, 1, vec![1.0]);
    let expected: DMatrix<f64> = DMatrix::from_vec(2, 1, vec![-0.5, 0.5]);
    let result = min_quad_w_equal(&P, &q, &r, &A, &b).unwrap();
    assert_eq!(expected, result);
}

#[test]
fn lhs() {
    let P: DMatrix<f64> = DMatrix::from_vec(2, 2, vec![1.0, 2.0, 3.0, 4.0]);
    let A: DMatrix<f64> = DMatrix::from_vec(3, 2, vec![5.0, 6.0, 7.0, 8.0, 9.0, 10.0]);
    let expected: DMatrix<f64> = DMatrix::from_vec(
        5,
        5,
        vec![
            1.0, 2.0, 5.0, 6.0, 7.0, 3.0, 4.0, 8.0, 9.0, 10.0, 5.0, 8.0, 0.0, 0.0, 0.0, 6.0, 9.0,
            0.0, 0.0, 0.0, 7.0, 10.0, 0.0, 0.0, 0.0,
        ],
    );
    let result = min_quad_w_equal_lhs(&P, &A);
    assert_eq!(expected, result);
}

#[test]
fn rhs() {
    let q: DMatrix<f64> = DMatrix::from_vec(3, 1, vec![13.0, 11.0, 7.0]);
    let b: DMatrix<f64> = DMatrix::from_vec(5, 1, vec![5.0, 17.0, 19.0, 23.0, 29.0]);
    let expected: DMatrix<f64> =
        DMatrix::from_vec(8, 1, vec![-13.0, -11.0, -7.0, 5.0, 17.0, 19.0, 23.0, 29.0]);
    let result = min_quad_w_equal_rhs(&q, &b);
    assert_eq!(expected, result);
}

File: optim.py
import numpy as np
from math import sqrt
from scipy import linalg


class OptimResult:
    def __init__(self, x, iterations, converged):
        self.x = (x,)
        self.iterations = (iterations,)
        self.converged = converged

    def __str__(self):
        return "\n".join(
            [
                f"x: {self.x}",
                f"iterations: {self.iterations}",
                f"converged: {self.converged}",
            ]
        )


def barrier(
    scaled_obj_w_log_barrier,
    scaled_obj_w_log_barrier_grad,
    scaled_obj_w_log_barrier_hess,
    t,
    mu,
    x,
    m,
    A,
    b,
    eps,
    max_iter,
):
    for iteration in range(max_iter):
        f = lambda x: scaled_obj_w_log_barrier(t, x)
        grad = lambda x: scaled_obj_w_log_barrier_grad(t, x)
        hess = lambda x: scaled_obj_w_log_barrier_hess(t, x)
        newton_result = newton_w_equal(f, grad, hess, x, A, b, eps, max_iter)
        if newton_result.converged:
            x = newton_result.x[0]
        else:
            return OptimResult(x, iteration, False)
        if m / t < eps:
            return OptimResult(x, iteration, True)
        t = mu * t
    return OptimResult(x, max_iter, False)


def newton_w_equal(
    f,
    grad,
    hess,
    x,
    A,
    b,
    eps,
    max_iter,
    backtrack=True,
    hessian_cutoff=True,
    cutoff=1e10,
):
    if A.shape[0] != A.shape[1]:
        raise Exception("A must be square")
    if linalg.norm((A @ x) - b) > 0.0001:
        raise Exception("Invalid initial guess")
    r = 0
    b = np.zeros((A.shape[0], 1))
    for iteration in range(max_iter):
        H = hess(x)
        if hessian_cutoff and linalg.norm(H) > cutoff:
            return OptimResult(x, iteration, True)
        G = grad(x)
        step = min_quad_w_equal(H, G, r, A, b)
        t = 1
        if backtrack:
            while f(x + t * step) >= f(x) and t > 0:
                t = 0.5 * t
        step = t * step
        dec = newton_dec(step, H)
        if (dec**2) / 2.0 < eps:
            return OptimResult(x, iteration, True)
        x = x + step
    return OptimResult(x, max_iter, False)


def newton_dec(h, H):
    return sqrt((h.T @ H @ h)[0, 0])


def min_quad_w_equal(P, q, r, A, b):
    lhs = min_quad_w_equal_lhs(P, A)
    rhs = min_quad_w_equal_rhs(q, b)
    try:
        sol = linalg.solve(lhs, rhs)
    except:
        sol = linalg.lstsq(lhs, rhs)[0]
    return sol[0 : A.shape[1]]


def min_quad_w_equal_lhs(P, A):
    return np.block([[P, A.T], [A, np.zeros((A.shape[0], A.shape[0]))]])


def min_quad_w_equal_rhs(q, b):
    return np.block([[-1 * q], [b]])


def unconstrained_newton(f, J, hess, x0, eps, max_iter):
    x = x0
    for iteration in range(max_iter):
        try:
            h = linalg.solve(hess(x), -1.0 * J(x))
        except ValueError:
            return OptimResult(x, iteration, True)
        x = x + h
        if linalg.norm(h) < eps:
            return OptimResult(x, iteration, True)
    return OptimResult(x, max_iter, False)
